crates/core/src/node/inner/mod.rs

```diff
@@ -0,0 +1,77 @@
+//! This module encapsulates mutable parts of the system and provides read-only views on various
+//! components of the system's state (e.g. time, storage, blocks). It is still possible to mutate
+//! the state outside of this module but only through [`InMemoryNodeInner`]'s public high-level
+//! methods.
+//!
+//! The idea behind this is being able to read current time to answer API requests while a lock on
+//! [`InMemoryNodeInner`] is being held for block production. At the same time it is impossible to
+//! advance the time without holding a lock to [`InMemoryNodeInner`].
+//!
+//! FIXME: The above is not 100% true yet (there are some internal parts of InMemoryNodeInner that
+//!        are available outside of this module)
+pub mod blockchain;
+pub mod fork;
+mod in_memory_inner;
+pub mod node_executor;
+pub mod time;
+
+pub use in_memory_inner::{InMemoryNodeInner, TxExecutionOutput};
+
+use crate::filters::EthFilters;
+use crate::node::{ImpersonationManager, TestNodeFeeInputProvider};
+use crate::system_contracts::SystemContracts;
+use anvil_zksync_config::constants::NON_FORK_FIRST_BLOCK_TIMESTAMP;
+use anvil_zksync_config::TestNodeConfig;
+use blockchain::BlockchainReader;
+use fork::{ForkDetails, ForkStorage};
+use std::sync::Arc;
+use time::TimeReader;
+use tokio::sync::RwLock;
+
+impl InMemoryNodeInner {
+    // TODO: Bake in Arc<RwLock<_>> into the struct itself
```

Comment:
Most of structures already have rwlocks inside, no? This indeed feels a bit clunky

Reply:
Yeah I left `Arc<RwLock<_>>` for now just in case but I suspect I can get rid of it at some point.